### ISSUE TITLE
gyro inverted over bt

### ISSUE THREAD

**Issue by rozhkovets** [Ryochan7/DS4Windows#929]
when ds4 is connected by bluetooth, the gyro is incorrect.
when turning left in the horizontal plane( Y axis?), the pad test and game in cemu, example zelda shows that the gamepad is turned right and vice versa.

when connected via usb everything is ok

ds4w v1.7.24
windows 10 1809
bt dongle with windows stack

**Comment by mika-n** [comment]
#767 

At this point it is unclear if this is a gamepad hardware bug or software bug in PC side. Maybe the only solution is to make a configuration tweak to UDPServer  (used by PadTest and Cemu) to optionally invert gyro axies read from a gamepad. This bug is an issue only with some users and gamepads, so difficult to say what is the root cause.

I'll try to make a tweak to debug version of DS4Windows to test this "invert UDP server reported gyro axis" trick. If it works then the change may make it to the official version also.

**Comment by mika-n** [comment]
By the way. If you map gyro as mouse in DS4Windows profile then is the result inverted in this case also? 
And how does the gyro movement look like in DS4Windows ControllerReadings (in profile editor) tab page?

Or is gyro inverted only while DS4 is used over BT in PadTest and Cemu apps?

**Comment by rozhkovets** [comment]
Yes, gyro inverted only while DS4 is used over BT.
Yes, If i maped gyro as mouse in DS4Windows profile then is the result inverted

make video with Padtest and DS4Windows ControllerReadings at movement https://yadi.sk/i/0SN2PngYVRPw0A

**Comment by rozhkovets** [comment]
When I used scptoolkit with the patch from the instruction https://cemuhook.sshnuke.net/padudpserver.html , everything worked well for both usb and bluetooth with default settings. 

Unfortunately, I can not use it again because of problems with installing the driver for bluetooth.

**Comment by mika-n** [comment]
@rozhkovets  Please try the new version of ds4Windows debug version (based on the latest version). There is now experimental support to invert yaw/roll/pitch gyro axis sensor values.
https://github.com/Ryochan7/DS4Windows/wiki/Troubleshooting#detailed-debug-log-file-from-gamepad-connection-problems

DS4WindowsDebug.exe.config configuration file has new Debug_GyroMultiplierYaw/Roll/Pitch options. By default the value is "1" meaning that the raw gyro sensor value (or value run through a calibration table) is use the way it comes from DS4 gamepad. But, if you set the Debug_GyroMultiplierxxx value to "-1" then the result would be inverted gyro output value.

Please try Debug_GyroMultiplierYaw with a value "-1". It should invert the "yaw" axis (left/right tilt). Changes in this configuration file takes effect after you restart DS4WindowsDebug.exe executable, so modify the file while the debug app is not running.

```
  <!-- Invert raw hardware SixAxis gyro yaw/roll/pitch values. Some DS4 gamepads have a bug and yaw may need invert logic -->
  <!--   1=Original gyro axis, -1=Invert gyro axis value -->
  <add key="Debug_GyroMultiplierYaw"   value="-1" />
  <add key="Debug_GyroMultiplierRoll"  value="1" />
  <add key="Debug_GyroMultiplierPitch" value="1" />
```

If this works then I have to think about how to properly integrate these options to DS4Windows app (profile or just global app specific config? Where to place these options in GUI?).

This "invert gyro axis" config option is just a workaround to the problem some users have. The best option would be to modify gyro calibration routine to somehow identify those gamepads needing this "invert logic". Someone with both working and "non-working inverted" gamepads should debug what is different between those DS4 hardware units.

**Comment by rozhkovets** [comment]
it works,
  <add key="Debug_GyroMultiplierYaw"   value="-1" />

for other axes this also works as intended
It works also when ds4 connected via usb. 

with modified exe.config and activated setting "Quick Charge" -
1)connect via bt - yaw is ok
2)connect via usb (example, need charge battery) - yaw is ok, bt was disabled
3)disconnect via usb, connect via bt - yaw is inverted, need restart ds4windows

**Comment by jinkazama82it** [comment]
a very strange issue XD i never encounter it but I have a hint, if the issue is generated when you disconnect the controller to the PC after completing the charge simply use an external charger for smartphone a 5v 1A charger is what you need.

**Comment by mika-n** [comment]
> it works, It works also when ds4 connected via usb.
> with modified exe.config and activated setting "Quick Charge" -

Hmm... Now I'm a bit confused because I understood that this "inverted yaw gyro axis" was a problem only when DS4 was connected using BT. Over USB connection the gyro worked correctly all the time and there was no need to reverse gyro axies?

When you set DS4WindowsDebug.exe.config Debug_GyroMultiplierXXX=-1 then it should reverse the gyro axis for USB connection also. Meaning that if the multiplier is -1 then USB connection should show inverted result (=wrong) because the originally correct orientation was unnecessarily reversed. GyroMultiiplier debug option should be set -1 only when gamepad is connected over BT (with these problematic gamepads).

> 1)connect via bt - yaw is ok
> 2)connect via usb (example, need charge battery) - yaw is ok, bt was disabled
> 3)disconnect via usb, connect via bt - yaw is inverted, need restart ds4windows

Hmm... Have you at first connected the gamepad over BT and then in the middle of the game session you plug in USB cable? Are you sure the connection type switched over to USB because it is possible to use BT connection and use usb cable only to charge the gamepad (ie. no data transmissions). I have several mobile phone usb cables which just charge the gamepad but doesn't carry over USB data from DS4 gamepads (don't know why. Some cable connection isssue).

**Comment by rozhkovets** [comment]
> Hmm... Now I'm a bit confused because I understood that this "inverted yaw gyro axis" was a problem only when DS4 was connected using BT. Over USB connection the gyro worked correctly all the time and there was no need to reverse gyro axies?

Yes, over USB connection the gyro worked correctly all the time. Now, Debug_GyroMultiplierXXX affects both types of connection. 

> Hmm... Have you at first connected the gamepad over BT and then in the middle of the game session you plug in USB cable? Are you sure the connection type switched over to USB because it is possible to use BT connection and use usb cable only to charge the gamepad (ie. no data transmissions).

Yes, in the status column, the icon changes to usb. 
Log file https://yadi.sk/d/gVO2pa3mwgS8Ng

**Comment by rozhkovets** [comment]
I noticed an bug, but so far I can not clearly repeat it. 
I launch DS4WindowsDebug with modified DS4WindowsDebug.exe.config, connecting ds4 via BT and the gyroscope is incorrect in the same way, as if DS4WindowsDebug.exe.config is not modified.

restart DS4WindowsDebug several times, run as administrato and as user, connect via usb/bt and finally everything comes back to normal.

**Comment by mika-n** [comment]
Hmm.... What is the content of Logs tab page when you launch the debug version, connect the gamepad with BT and still gyro is wrong even when the config option inverted it?  

I can't think of any other reason but sometimes gyro calibration works correctly and GyroMultipplierXXX=-1 is not needed, but because it is hard coded config option the result will be wrong. After several restarts the gyro calibration "fails" and the raw gyro orientation is inverted but in this case the config option fixes it in software side.

Maybe i need to add few more debug messages to a calibration routine to see are there differences between restarts (and between working and non-working raw gyro orientation). This is veeery strange. I'll let you know in this issue if/when there is a new debug version available with additional calibration debug messages. It may help us to track down the root cause.

**Comment by rozhkovets** [comment]
> Logs tab page when you launch the debug version, connect the gamepad with BT and still gyro is wrong even when the config option inverted it

log with disabled "Hide DS4 Controller"
https://yadi.sk/d/xmAUvUYuSyEk8A

log with enabled "Hide DS4 Controller"
https://yadi.sk/d/nz_Mh_cCCjztlQ

may be exclusive mode dont work
___________________________
I try install HidGuardian, it was hard and only manual installation worked
also, there is no 'Write HID Guardian Device List' item on the settings tab, run as administrator of course
https://yadi.sk/i/ApyNHaaV09y_lg
_________
nothing has changed, but I'm not sure that everything installed correctly
ProcessExplorer displayed what HID-game device uses only DS4Windows, virtual xbox360 gamepad uses only dwm.exe

**Comment by mika-n** [comment]
When debugging things then launch DS4Windows.exe or DS4WindowsDebug.exe from FileExplorer or desktop shortcut and not as "run at startup" or "run as task". For testing purposes try to disable (untick) "Run at Startup" option in DS4Windows settings and launch the DS4WindowsDebug.exe from FileExplorer.

There has been cases where some weird permission issue in "Run at Startup" has caused problems. Especially the use of "Hide DS4 Windows" option needs elevated priviledges to let DS4Windows app to hide the original physical HID controller.

"Write HID Guardian Device List" button is gone long time ago in DS4Windows. Some old instructions may still refer to it, but the button is no longer available. HIDGuardiaTestInstaller tool is the easiest way to install HidGuardian tool and setup sysreg entry AffectedDevices.
https://github.com/Ryochan7/DS4Windows/wiki/DS4Windows-and-HIDGuardian-Install-and-Setup-Guide#hidguardian-test-installer

**Comment by kishelraj** [comment]
I thought I was going crazy when I reported the issue a while back. Thank goodness I am not the only one with this issue.

**Comment by rozhkovets** [comment]
> try to disable (untick) "Run at Startup"

I enable it just to show what ds4w running as administrator, otherwise, the uac icon will be next to the checkmark "Task".

> HIDGuardiaTestInstaller
https://github.com/Ryochan7/DS4Windows/wiki/DS4Windows-and-HIDGuardian-Install-and-Setup-Guide#hidguardian-test-installer

Link removed. Need newer executable. 😅Okay

Manual install from this instruction led to error devcon.exe failed. 
I used 
`devcon.exe install HidGuardian.inf Root\HidGuardian
devcon.exe classfilter HIDClass upper -HidGuardian
devcon.exe classfilter XnaComposite upper -HidGuardian
devcon.exe classfilter XboxComposite upper -HidGuardian`
And manual add hardware ID in registry

**Comment by rozhkovets** [comment]
@kishelraj 
> I thought I was going crazy when I reported the issue a while back. Thank goodness I am not the only one with this issue.

I have a suspicion that my ds4 cuh-zct2u is not original, or a very good copy.  Of the differences that I found on the Internet, the battery (noname, in blue pvc) does not correspond and the text on the motherboard is not like the others.  Produced on 09/2019.  Can you check yours gamepad?

**Comment by mika-n** [comment]
> Link removed. Need newer executable. 😅Okay

Uuups. @Ryochan7 has probably taken the old installer offline. Too bad because it was "one click" process to install HidGuardian and to initialize AffectedDevices sysreg entry.

> Manual install from this instruction led to error devcon.exe failed.

Haven't tried the manual process. Did you run it in "Run as Administrator" cmd shell?

**Comment by rozhkovets** [comment]
> Haven't tried the manual process. Did you run it in "Run as Administrator" cmd shell?

Yes of course, script from HIDGuardian Helper Files checks for access rights and reports an error if they are not.

**Comment by mika-n** [comment]
Easy way to verify that HidGuardian is doing its magic is to open JOY.CPL controlPanel applet when you have connected the gamepad. If HidGuardian is doing the job then the original physical controller should not be listed there (with or without DS4Windows app running). If the original WirelessController is shown there then HidGuardian setup is not yet correct.

**Comment by Ryochan7** [comment]
Meant to get around to making a new installer last weekend. Ended up working on Touchmote changes instead. I actually found a very old source tree of the old installer on my laptop today; it was from an old VPS data backup from around 2018/02/27. It was from before that program had an uninstall routine. Some of the old code was still useful though.

I have ported the old ViGEmBus installer program to install HidGuardian instead. I have tested both the install and uninstall routines on my system. I did incorporate a couple of changes including attempting to remove the UpperFilters entry first during the uninstall. This time around the installer code will be pushed to GitHub so I don't lose progress again. It should be available fairly soon.

**Comment by Ryusennin** [comment]
For what it's worth, I had to deal with another Slow Ring update two days ago (version 2004, build 19041), and of course I had to reactivate HidGuardian. The test installer (quick uninstall/reinstall procedure) worked as usual without any complication.

Not that I spend a lot of time with Win10, but sometimes I don't have a choice :/

**Comment by Ryochan7** [comment]
I use a couple pieces of software that require running Windows 10 so I am pretty much stuck with Win10 on my main Windows partition; Docker for Windows is an example.

I got a simple HidGuardian installer made and archived now. It is a bit more minimal than the previous installer but it works well so far. This installer uses NLog for writing log files and it can catch some crash errors so hopefully an issue can be found easily if something goes wrong. The installer has been tested on Windows 10 and Windows 7.

HidGuardianInstaller_20191218.zip
https://drive.google.com/file/d/1PNL3uv_4KektN00S9fm61djypSQ-3HED/view?usp=sharing

Code:
https://github.com/Ryochan7/HidGuardianInstaller

**Comment by rozhkovets** [comment]
well, I install HidGuardian, gamepad not listed in joy.cpl 

> I noticed an bug, but so far I can not clearly repeat it.
> I launch DS4WindowsDebug with modified DS4WindowsDebug.exe.config, connecting ds4 via BT and the gyroscope is incorrect in the same way, as if DS4WindowsDebug.exe.config is not modified.

> with modified exe.config and activated setting "Quick Charge" -
> 1)connect via bt - yaw is ok
> 2)connect via usb (example, need charge battery) - yaw is ok, bt was disabled
> 3)disconnect via usb, connect via bt - yaw is inverted, need restart ds4windows


during the experiments I got this
1) If i connect gamepad via bt first, after i run ds4windows with with modified DS4WindowsDebug.exe.config 
` <add key="Debug_GyroMultiplierYaw"   value="-1" />`
then gyro is normal, settings worked.
after, reconnect gamepad via bt without stopping ds4windows - gyro is reversed, settings did not worked.
log file https://yadi.sk/d/vW-fz__NtEi2vA

2) If i run ds4windows with with modified DS4WindowsDebug.exe.config first,
after i connect gamepad via bt, then gyro is reversed, settings did not worked.
after, reconnect gamepad via bt without stopping ds4windows - gyro is reversed, settings did not worked.

log file https://yadi.sk/d/NLgxsjS3M5coCA

I think that Quick Charge not the source of the problem. The problem is in the order of launch and connection

**Comment by mika-n** [comment]
@rozhkovets  Please download yet-another-new-version of the debug DS4Windows app version. It should fix the issue of ignoring GyroMultiplierXXX config option when a controller is hotplugged in. It indeed used to work only when DS4Windows app was already running when a gamepad was connected (there are different initialization routes in "cold plug" and "hot plug" situtations).

Also, the new version dumps out additional gyro calibration debug values (DS4Windows Log tab page). I would be veeeery interested to see how those refreshCalibrationData and setCalibrationData debug lines look like with a gamepad having this inverted gyro axis issue is connected to PC. Please export the Logs tab content and provide a link to the TXT file here. One of those calibration values must be somehow weird compared to normal DS4 gamepads.

When you generate the debug content then please do it at first using USB connection and export the log content (ie. as a reference information while USB should not have the issue). Then disconnect the gamepad from PC and close DS4WindowsDebug.exe app, re-launch the debug app and try again with BT connection. Now there should be some differences in calibration data and it might explain why gyro is inverted in BT.

**Comment by rozhkovets** [comment]
checked, in yet-another-new-version "cold plug" and "hot plug" works without problems

USB https://yadi.sk/d/3OdjVUq0iJ683A
BT https://yadi.sk/d/VncTD3SqdwtXOw

**Comment by mika-n** [comment]
@rozhkovets  Thank you for those debug log files. I think we are on to something here. There is clearly a weird ratio between gyro and accelerator values in YAW axis. Is it so that only YAW axis is reversed in BT (ie. roll and pitch gyro axies are OK)? At least it looks like it based on the debugged calibration values.

USB with working gyro in yaw/pitch/roll:
```
20.12.2019 20:05:21: DEBUG: setCalibrationData. FromUsb=True
20.12.2019 20:05:21: DEBUG: setCalibrationData. calibrationData[0] bias=0  sensNumer=17280  sensDenom=17569
**20.12.2019 20:05:21: DEBUG: setCalibrationData. calibrationData[1] bias=0  sensNumer=17280  sensDenom=17569**
20.12.2019 20:05:21: DEBUG: setCalibrationData. calibrationData[2] bias=0  sensNumer=17280  sensDenom=17569
20.12.2019 20:05:21: DEBUG: setCalibrationData. calibrationData[3] bias=0  sensNumer=16384  sensDenom=16384
20.12.2019 20:05:21: DEBUG: setCalibrationData. calibrationData[4] bias=0  sensNumer=16384  sensDenom=16384
20.12.2019 20:05:21: DEBUG: setCalibrationData. calibrationData[5] bias=0  sensNumer=16384  sensDenom=16384
20.12.2019 20:05:21: DEBUG: setCalibrationData. gyroSpeed2x=1080 gyroSpeedPlus+gyroSpeedMinus=(540 + 540)  yaw=17569 yawPlus-yawMinus=(8829 - -8740)
```

BT with a gamepad sending reversed gyro/accel values in YAW axis (calibrationData[1]):
```
20.12.2019 20:06:44: DEBUG: setCalibrationData. FromUsb=False
20.12.2019 20:06:44: DEBUG: setCalibrationData. calibrationData[0] bias=0  sensNumer=17280  sensDenom=17569
**20.12.2019 20:06:44: DEBUG: setCalibrationData. calibrationData[1] bias=0  sensNumer=17280  sensDenom=-17569**
20.12.2019 20:06:44: DEBUG: setCalibrationData. calibrationData[2] bias=0  sensNumer=17280  sensDenom=17569
20.12.2019 20:06:44: DEBUG: setCalibrationData. calibrationData[3] bias=0  sensNumer=16384  sensDenom=16384
20.12.2019 20:06:44: DEBUG: setCalibrationData. calibrationData[4] bias=0  sensNumer=16384  sensDenom=16384
20.12.2019 20:06:44: DEBUG: setCalibrationData. calibrationData[5] bias=0  sensNumer=16384  sensDenom=16384
```

So, if the scenario is that only one gyro axis (YAW in your case) is reversed then I can make a code change to automagically identify that one exceptional axis calibration compared to other two axies. 
I think this is an issue only in some DS4 revision 1 gamepads, because I have not heard anyone using DS4 revision 2 gamepad complaining about reversed gyro. By doing this automagical identification and reversion then there would be no need to do config or profile option tweaks and changing a controller type on the fly between BT and USB would not cause problems.

**Comment by mika-n** [comment]
@rozhkovets  @kishelraj  Here is a test version of DS4WindowsWPF 1.9.x  app with an experimental automagical identification and fixup of inverted YAW gyro axis (ie. no need to tweak any config file options and the same config files work for both working and broken DS4 rev1 gyros).

However, please note that this "gyro_fixup" version is based on the latest beta version of DS4WindowsWPF x64 64bit build which will soon replace the DS4Windows WinForms version everyone have been using so far.

To use the linked DS4WindowsWPF_FixupInvertedYaw version linked here you have to at first download downloaded the latest DS4WindowsWPF beta version.

Step 1: Download DS4WindowsWPF x64 beta 1.9.2 version and extract it to a new folder (ie. don't replace files in your existing DS4Windows "WinForms" app folder). But old and this new version use the same config and profile files, so those will be shared in all app versions. 
https://github.com/Ryochan7/DS4Windows/issues/824#issuecomment-568109267

Step 2: Download DS4WindowsWPF_FixupInvertedYaw.zip and replace the DS4WindowsWPF executable from step 1 with this this "fixupYaw executable" version. 
[DS4WindowsWPF_FixupInvertedYaw.zip](https://github.com/Ryochan7/DS4Windows/files/3991752/DS4WindowsWPF_FixupInvertedYaw.zip)

Step 3: Launch the beta DS4Windows WPF version and test the problematic gamepad. You should see a message in Logs tab page if the gyro axis calibration was automatically fixed.

Please test gyro using Cemu/PadTest/DS4Windows gyro-to-mouse functionalities. All those features should now work OK in YAW gyro axis also. And if you see any other problems in this beta version then please report those also.

**Comment by rozhkovets** [comment]
> test gyro using Cemu/PadTest/DS4Windows gyro-to-mouse functionalities and plug/unplug/re-plug gamepad scenarios.
everything works great, the problem with the gyroscope can be considered solved.

in the new version of the program other language packs do not work, always "Englsih"
also, when activated "close minimizes" dont work middle mouse click (tray notification icon) for exit from app

**Comment by mika-n** [comment]
Sounds great. I'll create a code fix for the gyro problem, but it will be included in this new WPF app version because the old WinForms version no longer receive changes (at least actively).

Many of the old languages supported in WinForms version were not maintained for years. There were looooots of "holes" in translations, so WPF version starts from a clean table (after all the GUI is using different technique, so translations should have been re-done anyway). Ryochan7 takes care of English and Japanese languages. I'll take care of Finnish (and maybe Swedish) but the beta version of WPF is not yet there.

"Exit from app" tooltray menu command works here even when "Close minimizes" option is enabled. Althought I don't have a "middle mouse button" in the touchpad of my laptop, so don't know if there are some issues with a physical middle mouse button clicks if that's what you meant.

**Comment by Ryochan7** [comment]
Currently, only English and Japanese are supported languages. The new way of providing translations is a bit different than the old method. Part of the old translations used by the DS4Control package could still be used but the GUI translations would have to be remade. Many language packs were very incomplete anyway so I figured that it would be best to start from scratch. It is interesting to see screenshots when the non-English language packs are used and how wrong the formatting can become. The use of layouts in the WPF version should make it much less susceptible to layout issue when rendering different translation strings.

I cannot take it upon myself to re-translate the application to all the old supported languages. My understanding of Japanese is fairly minimal (only JLPT N5 level) but I was able to convert some of the old translation strings and add some new strings to the new Japanese language pack.

As for the tray icon, that should be an easy fix. It is just a conflict of two settings. The callback method just calls close on the main window so the "Close Minimizes" routine is activated. You would have to choose the explicit Exit option from the tray menu as there is an override flag in that case.
